File: OC/内容层动画/WGOCCoreAnimation/WGOCCoreAnimation/CAEmitterCell粒子动画/大火动画/WGFireAniViewController.m
//
//  WGFireAniViewController.m
//  WGOCCoreAnimation
//
//  Created by wanggang on 2019/5/23.
//  Copyright © 2019 bozhong. All rights reserved.
//

#import "WGFireAniViewController.h"

@interface WGFireAniViewController ()

@property (nonatomic, strong) CAEmitterLayer *emitterLayer;
@property (nonatomic, strong) CAEmitterCell *emitterCell;

@end

@implementation WGFireAniViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title = @"大火动画";
    self.view.backgroundColor = UIColor.blackColor;
    
    //粒子发射器
    self.emitterLayer = [CAEmitterLayer layer];
    //粒子发射位置
    _emitterLayer.position = self.view.center;
    //控制火苗大小
    _emitterLayer.emitterSize = CGSizeMake(SCREEN_WIDTH, 30);
    _emitterLayer.renderMode = kCAEmitterLayerAdditive;
    //控制发射源模式 即形状
    _emitterLayer.emitterMode = kCAEmitterLayerOutline;
    _emitterLayer.emitterShape = kCAEmitterLayerLine;
    [self.view.layer addSublayer:_emitterLayer];
    
    self.emitterCell = [[CAEmitterCell alloc] init];
    _emitterCell.name = @"fire";
    //发射方向
    _emitterCell.emissionLongitude = M_PI;
    //粒子速度 负数表明向上燃烧
    _emitterCell.velocity = -10;
    //粒子速度范围
    _emitterCell.velocityRange = 20;
    //周围发射角度
    _emitterCell.emissionRange = 1.5;
    //粒子y方向的加速度分量
    _emitterCell.yAcceleration = -100;
    //缩放比例 超大火苗
    _emitterCell.scaleSpeed = 1.2;
    //火苗颜色
    _emitterCell.color = [UIColor colorWithRed:0.8 green:0.4 blue:0.2 alpha:0.1].CGColor;
    _emitterCell.contents = (__bridge id _Nullable)([UIImage imageNamed:@"fire.png"].CGImage);
    
    _emitterLayer.emitterCells = @[_emitterCell];
    [_emitterLayer setValue:@(500) forKeyPath:@"emitterCells.fire.birthRate"];
    [_emitterLayer setValue:@(1) forKeyPath:@"emitterCells.fire.lifetime"];
    
}



@end

File: Swift/内容层动画/WGCoreAnimation/WGCoreAnimation/粒子动画/鬼火动画/WGFireAniViewController.swift
//
//  WGFireAniViewController.swift
//  WGCoreAnimation
//
//  Created by wanggang on 2019/4/25.
//  Copyright © 2019 wanggang. All rights reserved.
//

import UIKit

class WGFireAniViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = UIColor.black
        title = "火动画"
        
        //粒子发射器
        let emitterLay = CAEmitterLayer()
        //粒子发射位置
        emitterLay.position = self.view.center
        // 控制火苗大小
        emitterLay.emitterSize = CGSize(width: 100, height: 50)
        emitterLay.renderMode = .additive
        // 控制发射源模式 即形状
        emitterLay.emitterMode = .outline
        emitterLay.emitterShape = .line
        self.view.layer.addSublayer(emitterLay)
        
        let emitterCell = CAEmitterCell()
        emitterCell.name = "fire"
        //发射方向
        emitterCell.emissionLongitude = CGFloat(Double.pi)
        // 粒子速度 负数表明向上燃烧
        emitterCell.velocity = -100
        // 粒子速度范围
        emitterCell.velocityRange = 20
        //周围发射角度
        emitterCell.emissionRange = 2
        //粒子y方向的加速度分量
        emitterCell.yAcceleration = -200
        //缩放比例 超大火苗
        emitterCell.scaleSpeed = 0.5
        //火苗颜色
        emitterCell.color = UIColor(red: 0.8 ,green: 0.4 ,blue:0.2 ,alpha:0.1).cgColor
        emitterCell.contents = UIImage(named: "fire.png")!.cgImage
        
        emitterLay.emitterCells = [emitterCell];
        emitterLay.setValue(1000, forKeyPath: "emitterCells.fire.birthRate")
        emitterLay.setValue(1, forKeyPath: "emitterCells.fire.lifetime")
        
    }

}
